src/Toolchain.ml

```diff
@@ -0,0 +1,500 @@
+  let projectRoot = Fpath.ofString folder in
+  Js.Promise.all2
+    ( PackageManager.available ~root:projectRoot ~env
+    , Manifest.lookup projectRoot
+      |> okThen (fun pms ->
+             if pms = PackageManagerSet.empty then
+               Error "TODO: global toolchain"
+             else
+               Ok pms) )
+  |> Js.Promise.then_
+       (fun (availablePackageManagers, alreadyUsedPackageManagers) ->
+         let availablePackageManagers =
+           match availablePackageManagers with
+           | Ok x -> x |> PackageManagerSet.of_list
+           | Error msg ->
+             Js.log2 "Error during availablePackageManagers()" msg;
+             PackageManagerSet.empty
+         in
+         let alreadyUsedPackageManagers =
+           match alreadyUsedPackageManagers with
+           | Ok x -> x
+           | Error msg ->
+             Js.log2 "Error during alreadyUsedPackageManagers()" msg;
+             PackageManagerSet.empty
+         in
+         Js.log "availablePackageManagers";
+         PackageManagerSet.iter
+           (fun x -> x |> PackageManager.toString |> Js.log)
+           availablePackageManagers;
+         Js.log "possiblyUsed";
+         PackageManagerSet.iter
+           (fun x -> x |> PackageManager.toString |> Js.log)
+           alreadyUsedPackageManagers;
+         match
+           PackageManagerSet.inter availablePackageManagers
+             alreadyUsedPackageManagers
+           |> PackageManagerSet.elements
+         with
+         | [] -> (
+           Js.log "Will lookup toolchain from global env";
+           match PackageManager.ofName projectRoot "<global>" with
+           | Ok t -> PackageManager.make ~env ~t
+           | Error msg -> Error msg |> Js.Promise.resolve )
+         | [ obviousChoice ] ->
+           Js.log2 "Toolchain detected" (PackageManager.toString obviousChoice);
+           PackageManager.make ~env ~t:obviousChoice
+         | multipleChoices -> (
+           let config = Vscode.Workspace.getConfiguration "ocaml" in
+           match
+             ( config
+               |. Vscode.WorkspaceConfiguration.get "packageManager"
+               |> Js.Nullable.toOption
+             , config
+               |. Vscode.WorkspaceConfiguration.get "toolChainRoot"
+               |> Js.Nullable.toOption )
+           with
+           | Some name, Some root ->
+             PackageManager.specOfName ~env ~name ~root:(Fpath.ofString root)
+           | Some name, None ->
+             PackageManager.specOfName ~env ~name ~root:projectRoot
+           | _ ->
+             (Window.showQuickPick
+                ( multipleChoices
+                |> List.map (fun pm -> PackageManager.toName pm)
+                |> Array.of_list )
+                (Window.QuickPickOptions.make ~canPickMany:false
+                   ~placeHolder:
+                     "Which package manager would you like to manage the \
+                      toolchain?"
+                   ()) [@bs])
```

Comment:
Maybe I am missing something, but why do we need `[@bs]` attribute here?

Reply:
At times, debugging got hard and readability of the output helped. Adding [@bs] removed the currying wrapper and made the output more readable